### PR TITLE
Client side ratelimit for DNS Record Controller

### LINKED ISSUE #387: Enforce client side ratelimit for DNS Record Controller
**How to categorize this issue?**

/area robustness
/kind enhancement
/priority 3
/platform alicloud

**What would you like to be added**:
We need to enforce client side ratelimit for DNS Record Controller. We already encountered AliCloud API [ratelimit issue](https://github.com/gardener/external-dns-management/issues/71) before. [TokenBucketRateLimiter ](https://github.com/kubernetes/client-go/blob/master/util/flowcontrol/throttle.go#L63) could be a reference.
**Why is this needed**:

### PR DESCRIPTION
**Pull request by shaoyongfeng**
**How to categorize this PR?**
/area control-plane
/kind enhancement
/platform alicloud

**What this PR does / why we need it**:
We need to enforce client side ratelimit for DNS Record Controller. We already encountered AliCloud API ratelimit issue before. 
**Which issue(s) this PR fixes**:
Fixes #387

**Special notes for your reviewer**:

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: charts/gardener-extension-provider-alicloud/values.yaml [hunk 1]
```diff
@@ -26,6 +26,9 @@ controllers:
     concurrentSyncs: 5
   dnsrecord:
     concurrentSyncs: 5
+    providerClientQPS: 25
```
Comment (jia-jerry) [inline charts/gardener-extension-provider-alicloud/values.yaml:29]: Would you mind provide a link where AliCloud specify the DNS Api ratelimit and put it above `providerClientQPS` as a comment?

File: pkg/alicloud/client/client_dns.go [hunk 2]
```diff
@@ -21,17 +21,30 @@ import (
...
+	"sigs.k8s.io/controller-runtime/pkg/log"
 )
 
 const (
-	domainsCacheTTL = 1 * time.Hour
+	domainsCacheTTL     = 1 * time.Hour
+	rateLimiterCacheTTL = 1 * time.Hour
 )
 
+// RateLimiterWaitError is an error to be reported if waiting for a aliyun dns  rate limiter fails.
```
Comment (jia-jerry) [inline pkg/alicloud/client/client_dns.go:38]:
```suggestion
// RateLimiterWaitError is an error to be reported if waiting for a aliyun dns rate limiter fails.
```

File: pkg/alicloud/client/client_dns.go [hunk 3]
```diff
@@ -40,16 +53,38 @@ func (f *clientFactory) NewDNSClient(region, accessKeyID, accessKeySecret string
...
+
+	// Get f rate limiter from the cache, or create f new one if not present
+	var rateLimiter *rate.Limiter
+	if v, ok := f.rateLimiters.Get(accessKeyID); ok {
+		rateLimiter = v.(*rate.Limiter)
+	} else {
+		rateLimiter = rate.NewLimiter(f.limit, f.burst)
+	}
+	// Set should be called on every Get with cache.Expiring to refresh the TTL
+	f.rateLimiters.Set(accessKeyID, rateLimiter, rateLimiterCacheTTL)
```
Comment (jia-jerry) [inline pkg/alicloud/client/client_dns.go:81]: Not investigate the source code of `rateLimiter`. Does `rateLimiters.Get` refresh the key? If so,  maybe `	f.rateLimiters.Set(accessKeyID, rateLimiter, rateLimiterCacheTTL)` can be placed in the else block.

File: pkg/alicloud/client/client_dns.go [hunk 4]
```diff
@@ -233,6 +288,18 @@ func (d *dnsClient) deleteDomainRecord(id string) error {
 	return nil
 }
 
+func (c *dnsClient) waitForRoute53RateLimiter(ctx context.Context) error {
```
Comment (jia-jerry) [inline pkg/alicloud/client/client_dns.go:291]: `Route53` is for AWS. Please rename like `AliCloudDNS` or sort of.

File: pkg/alicloud/client/client_dns.go [hunk 5]
```diff
@@ -21,17 +21,30 @@ import (
...
+	Cause error
+}
+
+func (e *RateLimiterWaitError) Error() string {
+	return fmt.Sprintf("could not wait for client-side route53 rate limiter: %+v", e.Cause)
+}
+
 // NewDNSClient creates a new DNS client with given region, accessKeyID, and accessKeySecret.
 func (f *clientFactory) NewDNSClient(region, accessKeyID, accessKeySecret string) (DNS, error) {
 	client, err := alidns.NewClientWithAccessKey(region, accessKeyID, accessKeySecret)
```
Comment (jia-jerry) [inline pkg/alicloud/client/client_dns.go:50]: How about cache the client instead? Aliyun client is threadsafe.

File: pkg/alicloud/client/client_dns.go (hunk 3 above)
Comment (shaoyongfeng) [reply pkg/alicloud/client/client_dns.go:81]: No, Get doesn't refresh the key but just return value of unexpired key

File: pkg/alicloud/client/client_dns.go (hunk 5 above)
Comment (shaoyongfeng) [reply pkg/alicloud/client/client_dns.go:50]: Currently both domain cache and ratelimiter cache are on factory level which benefits all dnsclient.  It's not so necessary to add another cache level.  From other side, if we cache dnsclient, the Get method of ratelimiter is never called anymore and TTL won't be refreshed.
So I don't suggest to do so now

File: charts/gardener-extension-provider-alicloud/values.yaml (hunk 1 above)
Comment (shaoyongfeng) [reply charts/gardener-extension-provider-alicloud/values.yaml:29]: Unfortunately there is no official document for it.  Ali supports said maximum QPS is 60

File: pkg/controller/dnsrecord/actuator.go [hunk 6]
```diff
@@ -168,15 +156,12 @@ func (a *actuator) getDomainName(ctx context.Context, dns *extensionsv1alpha1.DN
...
-				Cause:        fmt.Errorf("could not get DNS domain names: %+v", err),
-				RequeueAfter: requeueAfterOnProviderError,
-			}
+			return "", fmt.Errorf("could not get DNS domain names: %+v", err)
 		}
 		a.logger.Info("Got DNS domain names", "domainNames", domainNames, "dnsrecord", kutil.ObjectName(dns))
 		domainName := dnsrecord.FindZoneForName(domainNames, dns.Spec.Name)
 		if domainName == "" {
-			return "", fmt.Errorf("could not find DNS domain name for name %s", dns.Spec.Name)
+			return "", wrapAliClientError(err, fmt.Sprintf("could not find DNS domain name for name %s", dns.Spec.Name))
```
Comment (jia-jerry) [inline pkg/controller/dnsrecord/actuator.go:164]: why need to wrapAliClientError?

Comment (shaoyongfeng) [reply pkg/controller/dnsrecord/actuator.go:164]: In order to re-enqueue the event if error is caused by Throttling

Comment (jia-jerry) [reply pkg/controller/dnsrecord/actuator.go:164]: `err` is alway nil. No wrap is needed.

File: pkg/controller/dnsrecord/actuator.go [hunk 7]
```diff
@@ -168,15 +156,12 @@ func (a *actuator) getDomainName(ctx context.Context, dns *extensionsv1alpha1.DN
 		// getting all domain names of the account and searching for the longest domain name that is a suffix of dns.spec.Name
 		domainNames, err := dnsClient.GetDomainNames(ctx)
 		if err != nil {
-			return "", &reconcilerutils.RequeueAfterError{
-				Cause:        fmt.Errorf("could not get DNS domain names: %+v", err),
-				RequeueAfter: requeueAfterOnProviderError,
-			}
+			return "", fmt.Errorf("could not get DNS domain names: %+v", err)
```
Comment (jia-jerry) [inline pkg/controller/dnsrecord/actuator.go:159]: I think you need to `wrapAliClientError` here.

**Comment by jia-jerry** [comment]
@EmoinLanyu  need second opinion.

**Review by EmoinLanyu — APPROVED** [review]
/lgtm

File: pkg/alicloud/client/client_dns.go [hunk 8]
```diff
@@ -21,17 +21,30 @@ import (
...
 )
 
+// RateLimiterWaitError is an error to be reported if waiting for a aliyun dns rate limiter fails.
+// This can only happen if the wait time would exceed the configured wait timeout.
+type RateLimiterWaitError struct {
+	Cause error
+}
+
+func (e *RateLimiterWaitError) Error() string {
+	return fmt.Sprintf("could not wait for client-side route53 rate limiter: %+v", e.Cause)
```
Comment (EmoinLanyu) [inline pkg/alicloud/client/client_dns.go:45]:
```suggestion
	return fmt.Sprintf("could not wait for client-side alicloud dns rate limiter: %+v", e.Cause)
```

**Review by jia-jerry — APPROVED** [review]

**Pull request merged**